2009-12-14  Jakub Jelinek  <jakub@redhat.com>

	PR c++/41183
	* cp-tree.h (current_class_ptr): Give NULL even when cfun
	has NULL cfun->language.

	* g++.dg/torture/pr41183.C: New test.

--- gcc/cp/cp-tree.h.jj	2009-09-24 10:21:50.000000000 +0200
+++ gcc/cp/cp-tree.h	2009-12-14 15:06:38.000000000 +0100
@@ -883,7 +883,8 @@ struct language_function GTY(())
    expression for `*this'.  */
 
 #define current_class_ptr \
-  (cfun ? cp_function_chain->x_current_class_ptr : NULL_TREE)
+  (cfun && cp_function_chain					\
+   ? cp_function_chain->x_current_class_ptr : NULL_TREE)
 #define current_class_ref \
   (cfun ? cp_function_chain->x_current_class_ref : NULL_TREE)
 
--- gcc/testsuite/g++.dg/torture/pr41183.C.jj	2009-12-14 15:10:57.000000000 +0100
+++ gcc/testsuite/g++.dg/torture/pr41183.C	2009-12-14 15:11:01.000000000 +0100
@@ -0,0 +1,30 @@
+// PR c++/41183
+// { dg-do compile }
+
+void foo (const char *);
+
+template <int *>
+struct A
+{
+  template <typename T> A (const int &, T);
+  int i;
+};
+
+template <int *X>
+template <typename T>
+A<X>::A (const int &j, T) : i(j)
+{
+  foo (0);
+  foo (0);
+  foo (__PRETTY_FUNCTION__);
+}
+
+int N;
+
+struct B
+{
+  B ();
+  A<&N> a;
+};
+
+B::B() : a(N, 0) {}
